src/ai/backend/manager/server.py

```diff
@@ -454,6 +454,84 @@ async def monitoring_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
         await ectx.cleanup()
 
 
+@actxmgr
+async def hanging_session_managing_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+    from contextlib import suppress
+    from datetime import timedelta
+    from itertools import chain
+    from typing import Optional, Tuple
+
+    import sqlalchemy as sa
+    from dateutil.tz import tzutc
+
+    from .models import kernels
+    from .models.kernel import KernelStatus
+    from .models.utils import ExtendedAsyncSAEngine
+
+    async def _fetch_kernels_with_status_and_period(
+        db: ExtendedAsyncSAEngine,
+        status: Optional[KernelStatus] = None,
+        period: Optional[timedelta] = None,
+    ) -> Tuple[Tuple[str, str, str], ...]:
+        async with db.begin_readonly() as conn:
+            query = sa.select(kernels)
+            if status:
+                query = query.where(kernels.c.status == status)
+                if period is not None:
+                    query = query.where(
+                        (
+                            datetime.now(tz=tzutc())
+                            - kernels.c.status_history[status.name].astext.cast(
+                                sa.types.DateTime(timezone=True)
+                            )
+                        )
+                        > period
+                    )
+                query = query.order_by(kernels.c.status_history[status.name].asc())
+            result = await conn.execute(query)
+            return tuple(kernel.session_id for kernel in result.fetchall())
+
+    async def _terminate_hanging_sessions(interval: float = 30.0):
+        while True:
+            preparing_kernels = await _fetch_kernels_with_status_and_period(
+                root_ctx.db, status=KernelStatus.PREPARING, period=timedelta(minutes=30)
+            )
+            log.debug(f"{len(preparing_kernels)} PREPARING kernels found.")
+
+            terminating_kernels = await _fetch_kernels_with_status_and_period(
+                root_ctx.db, status=KernelStatus.TERMINATING, period=timedelta(minutes=30)
+            )
+            log.debug(f"{len(terminating_kernels)} TERMINATING kernels found.")
+
+            _ = await asyncio.gather(
+                *[
+                    asyncio.create_task(
+                        root_ctx.registry.destroy_session(
+                            functools.partial(
+                                root_ctx.registry.get_session_by_session_id,
+                                session_id,
+                            ),
+                            forced=True,
+                            reason="automatic-force-termination",
```

Comment:
How about distinguish the termination reason for the `PREPARING` and `TERMINATING` sessions? This will help to easily know in which status the system is failing.

Reply:
Thank you for the feedback! Now it provides a distinguishable reason for each preparing/terminating session. Please check the updated code below:
https://github.com/lablup/backend.ai/blob/9470b6907659124b8fbdeb3535d93cff8173a805/src/ai/backend/manager/server.py#L508-L521